src/kudu/consensus/consensus_meta.cc

```diff
@@ -83,18 +83,38 @@ void ConsensusMetadata::clear_voted_for() {
 
 void ConsensusMetadata::populate_previous_vote_history(
     const PreviousVotePB& prev_vote) {
-  previous_vote_history_.emplace(prev_vote.election_term(), prev_vote);
-
-  // Prune vote history, if necessary. It is done only when the vote history
-  // size goes beyond a certain limit and the oldest term being pruned is
-  // older than the last known leader by this node.
-  const std::map<int64_t, PreviousVotePB>::const_iterator it =
-      previous_vote_history_.begin();
-  if (previous_vote_history_.size() > VOTE_HISTORY_MAX_SIZE &&
-      last_known_leader_.election_term() > it->first) {
-    last_pruned_term_ = it->first;
-    VLOG_WITH_PREFIX(2) << "Pruning history older than: " << last_pruned_term_;
-    previous_vote_history_.erase(it);
+  google::protobuf::Map<int64_t, PreviousVotePB>* previous_vote_history =
+      pb_.mutable_previous_vote_history();
+  InsertIfNotPresent(
+      previous_vote_history, prev_vote.election_term(), prev_vote);
+
+  int64_t last_known_leader_term = pb_.last_known_leader().election_term();
+  int64_t last_pruned_term = pb_.last_pruned_term();
+
+  // Prune vote history, if necessary.
+  // Step 1: Prune all the way until last known leader's term.
+  google::protobuf::Map<int64_t, PreviousVotePB>::iterator begin_it =
+      previous_vote_history->begin();
+  google::protobuf::Map<int64_t, PreviousVotePB>::iterator end_it =
+      begin_it;
+  while (end_it != previous_vote_history->end() &&
+      end_it->first <= last_known_leader_term) {
```

Comment:
you could use the function upper_bound for this.

Reply:
Yeah, I was thinking the same but I haven't checked if https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.map#Map.iterator would work properly with standard library functions. I think it should but I guess doing a log(n) operation here only to do an o(n) delete later should not be much of a difference. I can revisit this later.